fork_choice/confirmation-rule.md

```diff
@@ -0,0 +1,639 @@
+    #
+    # If it is satisfied, then it means that ffg_support_for_checkpoint can be as high as the maximum fraction of
+    # adversarial stake that the consensus protocol can deal with, i.e., < 1/3.
+    # If it not satisfied, then we proceed to the next case.
+
+    if remaining_ffg_weight > 0:
+        ffg_confirmation_score = (
+            100 * (
+                3 * (ffg_support_for_checkpoint - min_ffg_support_slash_th + remaining_ffg_weight) 
+                - 2 * total_active_balance
+            ) // (3 * remaining_ffg_weight)
+        )
+        if min_ffg_support_slash_th * 100 <= total_active_balance * ffg_confirmation_score:
+            return ffg_confirmation_score
+    else:
+        if 2 * total_active_balance <= (ffg_support_for_checkpoint - min_ffg_support_slash_th) * 3:
+            return 100 // 3
+
+    # Case 2: min_ffg_support_slash_th > total_active_balance * ffg_confirmation_score / 100
+    #
+    # In this case the problem reduces to returning the max possible ffg_confirmation_score such that:
+    #
+    # 2 / 3 * total_active_balance <= \
+    #     ffg_support_for_checkpoint - \
+    #     total_active_balance * ffg_confirmation_score / 100 + \
+    #     (1 - ffg_confirmation_score / 100) * remaining_ffg_weight
+
+    ffg_confirmation_score = (
+        100 * (3 * (ffg_support_for_checkpoint + remaining_ffg_weight) - 2 * total_active_balance) //
+        (3 * (remaining_ffg_weight + total_active_balance))
+    )
+    assert ffg_support_for_checkpoint * 100 > total_active_balance * ffg_confirmation_score
+    return max(ffg_confirmation_score, UNCONFIRMED_SCORE)
+```
+
+#### `get_ffg_confirmation_score`
+
+```python
+def get_ffg_confirmation_score(store: Store, block_root: Root) -> int:
+
+    current_epoch = get_current_store_epoch(store)
+
+    block = store.blocks[block_root]
+    block_epoch = compute_epoch_at_slot(block.slot)
+
+    # This function is only applicable to current and previous epoch blocks
+    assert current_epoch in [block_epoch, block_epoch + 1]
+
+    total_active_balance = int(get_total_active_balance_for_block_root(store, block_root))
+
+    ffg_support_for_checkpoint = int(get_ffg_support(store, block_root))
+
+    min_ffg_support_slash_th = min(ffg_support_for_checkpoint, CONFIRMATION_SLASHING_THRESHOLD)
+
+    if block_epoch == current_epoch:
+        remaining_ffg_weight = int(get_remaining_weight_in_current_epoch(store, block_root))
+    else:
+        remaining_ffg_weight = 0  
+
+    return get_ffg_confirmation_score_helper(
+        total_active_balance,
+        ffg_support_for_checkpoint,
+        min_ffg_support_slash_th,
+        remaining_ffg_weight
+    )          
+```
+
+### `get_confirmation_score`
+
+```python
+def get_confirmation_score(store: Store, block_root: Root) -> int:
+    """
+    Return `UNCONFIRMED_SCORE` in the case that `block_root` cannot be confirmed even by assuming no adversary weight,
```

Comment:
Considering no adversary in the system, does it mean that such a block can be reorged even if the confirmation rule synchrony assumption holds?

Reply:
Yep. Essentially, because we assume that synchrony only holds from the current slot voting time onwards.

Take the following case:
1. No adversary
3. B is from the previous slot

Because of previous asynchrony, it could be that
1. the proposer for the current slot proposed a block B' conflicting with B as it had not received B by the time of proposing
2. No validator received B by the time they had to vote in the previous epoch.
2. All validators received B' by the time they had to vote in the current epoch. Due to proposer boost, then, they all LMD voted for B'. 

Hence, B' has more LMD weight than B.

In this case, B is being re-orged out, so B confirmation score would be -1

Reply:
I can see that. So it mainly depends on when the confirmation score logic is invoked. Synchrony starting from current slot would mean that all messages are received 8 seconds into the current slot, but if the logic is run at the beginning of current slot it won't consider all past messages and can confirm B (if enough votes have been received by a local node while other haven't seen those votes at all). It makes me think that the right moment to run the confirmation logic is at the end of a slot where we assume synchrony was restored -- this would also consider votes made in current slot in case of block B and B' and is important consideration for client's logic regardless of the case we're discussing here. 

Otherwise, if a client runs this logic at the beginning of a slot it can unconfirm already confirmed block because a new slot event adds to W_p while not adding to S_p yet.